File: php-zigar/demos/sqlite/zig/async.zig
const std = @import("std");

const zigar = @import("zigar");

const worker = @import("search.zig");
pub const @"meta(zigar)" = worker.@"meta(zigar)";

var work_queue: zigar.thread.WorkQueue(worker) = .{};

pub const search = work_queue.promisify(worker.search);
pub const startup = work_queue.promisify(.startup);
pub const shutdown = work_queue.promisify(.shutdown);

File: php-zigar/demos/sqlite/src/async.php
<?php

// $m = zigar_use(__DIR__ . '/../zig/async.zig');

// $path = __DIR__ . '/../chinook.db';
// $keyword = $_GET['q'] ?? '';
// $result = $m->search($path, $keyword);
// header('Content-Type: application/json; charset=utf-8');
// echo json_encode($result, JSON_PRETTY_PRINT);

// $m->shutdown();

require __DIR__ . '/../vendor/autoload.php';

use Revolt\EventLoop;

ini_set('zigar.event_loop', 'revolt');

EventLoop::defer(function (): void {
    $m = zigar_use(__DIR__ . '/../zig/async.zig');

    $path = __DIR__ . '/../chinook.db';
    $keyword = $_GET['q'] ?? '';
    $result = $m->search($path, $keyword);
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode($result, JSON_PRETTY_PRINT);

    $m->shutdown();
});
EventLoop::defer(function (): void {
    echo "[doing something else]\n";
});
EventLoop::run();